core/src/main/java/org/apache/calcite/rel/metadata/ReflectiveRelMetadataProvider.java

```diff
@@ -78,7 +78,9 @@ protected ReflectiveRelMetadataProvider(
       ConcurrentMap<Class<RelNode>, UnboundMetadata> map,
       Class<? extends Metadata> metadataClass0,
       Multimap<Method, MetadataHandler> handlerMap) {
-    assert !map.isEmpty() : "are your methods named wrong?";
+    if (map.isEmpty()) {
+      throw new IllegalArgumentException("ReflectiveRelMetadataProvider methods map is empty");
+    }
```

Comment:
There are many codes in Calcite use assert, should we also eliminate them ?

Reply:
No, I think generally assertions are a valid mechanism, but in this particular case in my opinion an exception is safer, for the reasons I mention in the Jira ticket: if assertions are deactivated we will not get later a runtime exception in this scenario, we will just end up silently constructing an invalid, empty ReflectiveRelMetadataProvider; and there is no mechanism in the caller module to verify this situation on its side. Hence this specific change to avoid this situation in 100% of the cases.

Reply:
`No, I think generally assertions are a valid mechanism` I'm not convinced by this, we may need to reorganize our assertion usage. Can you log an issue there ?

Reply:
Ok, @danny0405 . I will merge the current PR and open a separate ticket to discuss / review the general usage of assertions in the code